### PR TITLE
feat: missed stops endpoint

### PR DESCRIPTION
**Pull request by lemald**
Asana ticket: [Missed stops API endpoint](https://app.asana.com/0/1148853526253420/1206680523793247/f)

### CODE REVIEW

File: lib/skate_web/controllers/detour_missed_stops_controller.ex [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    shape_with_stops_fn =
+      Application.get_env(
+        :skate_web,
+        :shape_with_stops_fn,
+        &Schedule.shape_with_stops_for_trip/1
+      )
+
+    missed_stops_fn =
+      Application.get_env(:skate_web, :missed_stops_fn, &MissedStops.missed_stops/1)
```
Comment (firestack) [inline lib/skate_web/controllers/detour_missed_stops_controller.ex:24]: nit/suggestion(non-blocking): I have a little trouble focusing on the content of the function itself when having to also try and mentally ignore these setup env calls.
How would you feel about doing something similar to what I did in the `auth_controller` and instead define these as private functions?
https://github.com/mbta/skate/blob/a3c3e2495de9d27ee618dcb0db9359eda623af04/lib/skate_web/controllers/auth_controller.ex#L85-L90

I could also see `Mox` being a solution here? But I can't say I really understand it well enough yet to know how it'd be used.

Comment (lemald) [reply lib/skate_web/controllers/detour_missed_stops_controller.ex]: What do you think of 28f1c160?

Comment (firestack) [reply lib/skate_web/controllers/detour_missed_stops_controller.ex]: perfect! Thank you!

File: lib/skate_web/controllers/detour_missed_stops_controller.ex [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  defp shape_with_stops(trip_id) do
+    Application.get_env(
+      :skate_web,
+      :shape_with_stops_fn,
+      &Schedule.shape_with_stops_for_trip/1
+    ).(trip_id)
+  end
+
+  defp missed_stops(cfg) do
+    Application.get_env(:skate_web, :missed_stops_fn, &MissedStops.missed_stops/1).(cfg)
```
Comment (joshlarson) [inline lib/skate_web/controllers/detour_missed_stops_controller.ex:50]:
```suggestion
  defp missed_stops(args) do
    Application.get_env(:skate_web, :missed_stops_fn, &MissedStops.missed_stops/1).(args)
```

**Nit**: I got confused at first and thought that `cfg` was some kind of application-level config.

Comment (lemald) [reply lib/skate_web/controllers/detour_missed_stops_controller.ex:50]: I borrowed the `cfg` terminology from the `MissedStops` module itself, though I agree that `args` is probably clearer. The `MissedStops` struct is a little weird IMO because it's only used as the argument to the `missed_stops` function but maybe @firestack could speak to why it's implemented that way?

Comment (firestack) [reply lib/skate_web/controllers/detour_missed_stops_controller.ex:50]: I probably should've used something that would accept either a keyword list or map, but  I didn't want to make it overly complex. I was finding that it was easy to get the args confused and it ended up being difficult to understand at a glance when reading the code back, so for readability's sake, I went with a struct. 

I'm fine with changing it, though I think the arguments should continue to be named in the calling code. I liked using a struct because it allows for requiring all fields to be present and making that somewhat visible at "writing time" and compile time.

Comment (lemald) [reply lib/skate_web/controllers/detour_missed_stops_controller.ex:50]: My perhaps controversial opinion is that I'd just write it as a `/4` function, though I think we should hold off on any refactoring on that and do it in another ticket / PR if we want to. In the meantime I like @joshlarson's naming change so I'll apply the suggestion.

File: test/skate_web/controllers/detour_missed_stops_controller_test.exs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      missed_stop = Enum.at(shape_with_stops.stops, 0)
+
+      connection_start = shape_with_stops.points |> Enum.at(0) |> Util.Location.as_location!()
+      connection_end = shape_with_stops.points |> Enum.at(1) |> Util.Location.as_location!()
+
+      reassign_env(:skate_web, :route_pattern_fn, fn _ -> route_pattern end)
+      reassign_env(:skate_web, :shape_with_stops_fn, fn _ -> shape_with_stops end)
+
+      reassign_env(:skate_web, :missed_stops_fn, fn %MissedStops{
+                                                      connection_start: ^connection_start,
```
Comment (joshlarson) [inline test/skate_web/controllers/detour_missed_stops_controller_test.exs:23]: How come you mocked the algorithm, rather than just providing some values for `route_pattern` and `shape_with_stops`, etc, that have simple enough expected results?

I'm partly asking because mocking out the internal function makes this test kind of brittle, and makes the whole PR a bit tougher to review. If we update `missed_stops_fn` to return data in a different way, then this test will still pass, and won't tell us anything about how well this controller works in context.

Comment (lemald) [reply test/skate_web/controllers/detour_missed_stops_controller_test.exs:23]: See if this was TypeScript I'd just make sure that it was set up such that the mocked function has to have the same type signature as the original, meaning that any changes to the interface without updating the mocked function would trigger a type error...

In any event I'm not sure we could elegantly do that in Elixir / Dialyzer except for _maybe_ via behaviours. My concern with leaving the test to call `missed_stops/1` directly is that it's a very complicated function and mocking out input data and expected output data would involve a lot of geospatial reasoning that seems pretty disconnected from the controller itself. I think overall the current approach is the lesser evil given that the return type of the function (list of stops) seems like it will be fairly stable.

Comment (joshlarson) [reply test/skate_web/controllers/detour_missed_stops_controller_test.exs:23]: It's not exactly blocking my review of this (it does make reviewing it a bit more complicated, because I have to cross-reference this mock with the behavior of `missed_stops`), but at some point, would you object if I tried refactoring this test to use "realistic" test data instead? Getting reasonable fake data that abstracts away a lot of the geospatial reasoning is a good use case for factories, and we're starting to build the beginnings of a good factory system.

Basically, I'm not as convinced that this is the lesser of two evils, but I don't think it's worth blocking this PR on, and that's kind of a philosophical point anyway.

**Review by joshlarson — APPROVED** [review]

File: test/skate_web/controllers/detour_missed_stops_controller_test.exs (hunk 3 above)
Comment (lemald) [reply test/skate_web/controllers/detour_missed_stops_controller_test.exs:23]: Yeah I'd certainly check out the result if you want to try that!
